gdx/src/com/badlogic/gdx/math/Quaternion.java

```diff
@@ -160,40 +160,40 @@ public int getGimbalPole () {
 	/** Get the roll euler angle in radians, which is the rotation around the z axis. Requires that this quaternion is normalized.
 	 * @return the rotation around the z axis in radians (between -PI and +PI) */
 	public float getRollRad () {
-		final int pole = getGimbalPole();
-		return pole == 0 ? MathUtils.atan2(2f * (w * z + y * x), 1f - 2f * (x * x + z * z))
-			: (float)pole * 2f * MathUtils.atan2(y, w);
+		final double pole = getGimbalPole();
+		return (float)(pole == 0 ? Math.atan2(2.0 * (w * z + y * x), 1.0 - 2.0 * (x * x + z * z))
+			: pole * 2.0 * Math.atan2(y, w));
 	}
 
 	/** Get the roll euler angle in degrees, which is the rotation around the z axis. Requires that this quaternion is normalized.
 	 * @return the rotation around the z axis in degrees (between -180 and +180) */
 	public float getRoll () {
-		return getRollRad() * MathUtils.radiansToDegrees;
+		return (float)Math.toDegrees(getRollRad());
 	}
 
 	/** Get the pitch euler angle in radians, which is the rotation around the x axis. Requires that this quaternion is normalized.
 	 * @return the rotation around the x axis in radians (between -(PI/2) and +(PI/2)) */
 	public float getPitchRad () {
 		final int pole = getGimbalPole();
-		return pole == 0 ? (float)Math.asin(MathUtils.clamp(2f * (w * x - z * y), -1f, 1f)) : (float)pole * MathUtils.PI * 0.5f;
+		return (float)(pole == 0 ? Math.asin(MathUtils.clamp(2.0 * (w * x - z * y), -1.0, 1.0)) : pole * MathUtils.PI * 0.5);
 	}
 
 	/** Get the pitch euler angle in degrees, which is the rotation around the x axis. Requires that this quaternion is normalized.
 	 * @return the rotation around the x axis in degrees (between -90 and +90) */
 	public float getPitch () {
-		return getPitchRad() * MathUtils.radiansToDegrees;
+		return (float)Math.toDegrees(getPitchRad());
 	}
 
 	/** Get the yaw euler angle in radians, which is the rotation around the y axis. Requires that this quaternion is normalized.
 	 * @return the rotation around the y axis in radians (between -PI and +PI) */
 	public float getYawRad () {
-		return getGimbalPole() == 0 ? MathUtils.atan2(2f * (y * w + x * z), 1f - 2f * (y * y + x * x)) : 0f;
+		return (float)(getGimbalPole() == 0 ? Math.atan2(2.0 * (y * w + x * z), 1.0 - 2.0 * (y * y + x * x)) : 0.0);
```

Comment:
I'm wondering now how much of the round-trip error is from `MathUtils.atan2()`, rather from float/double precision. I think I can find or remake some old spinning-globe GIFs that had significant distortion due to a bad `atan2()` implementation (I don't know if it was the same as MathUtils, but I think it was). The MathUtils docs for `atan2()` say that the average error is 0.1323 degrees. I've been using a rather-different implementation for `atan2()` in my code lately, and even though I know its average and max error are better, I don't know how much better (might not be enough here). It's also slightly faster, last I benchmarked.

Reply:
Running my test against current implementation:
`expected:<-5.0> but was:<-5.0020123>`

Changing the `atan2()` usage results in:
`expected:<-5.0> but was:<-5.0000014>`

Switching to `Math.atan2()` would be a much smaller change and improve precision by several orders.

Reply:
Woah, that is a significant improvement. That means the error from the entire rest of the operation is only about 1.4e-6, including the other approximations (like `MathUtils.asin()`, which I know isn't as accurate as `Math.asin()`, but isn't very bad). I know `Math.atan2()` usually isn't as fast as an approximation, particularly on Java 8, but it does get optimized pretty well on Java 11 and newer. The actual error here is almost as small as it can get; -5.0000014 has the bit representation 0xC0A00003, and we want -5.0, or 0xC0A00000 -- this is a difference only in the two least significant bits. This could still accumulate, but would do so very, very slowly in most cases. The error before, with the current `Quaternion` and -5.0020123, is 0xC0A0107C, which is wrong in its 13 least significant bits.